### ISSUE TITLE
When comparing v4 to v3, significant time difference in the post run process

### ISSUE THREAD

**Issue by kdkdhoho** [actions/setup-java#591]
**Description:**
When comparing v4 to v3, we noticed a significant time difference in the post run process.

**Task version:**
Photo reference

**Platform:**
- [x] Ubuntu (actions)
- [x] macOS (local)
- [ ] Windows

**Runner type:**
- [x] Hosted
- [ ] Self-hosted

**Repro steps:**  
Below are the results using v4.
[Result Page of v4](https://github.com/8-Sprinters/ListyWave-back/actions/runs/7753323010)
<img width="1288" alt="image" src="https://github.com/actions/setup-java/assets/66300965/9470b39b-a80a-451c-994b-c0dbcf3f89ad">
<img width="424" alt="image" src="https://github.com/actions/setup-java/assets/66300965/ef793f20-9e19-41cc-9048-955b086d3fcb">

Below are the results using v3.
[Result Page of v3](https://github.com/8-Sprinters/ListyWave-back/actions/runs/7753459097/job/21144808813)
<img width="1294" alt="image" src="https://github.com/actions/setup-java/assets/66300965/971c24b1-2b9d-4e34-81da-af67e8aa32b3">
<img width="425" alt="image" src="https://github.com/actions/setup-java/assets/66300965/c32efdcf-ad9a-4126-9d8a-a3a87d072bf6">

**Expected behavior:**
There is no time difference between the two.

**Actual behavior:**
There was a significant time difference.

**Comment by brenoavm** [comment]
I'm having the same issue when running the following actions:

```
jobs:
  build-and-deploy:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout Repository
      uses: actions/checkout@v4

    - name: Set up JDK 17
      uses: actions/setup-java@v3
      with:
        java-version: "17"
        distribution: "oracle"
        cache: "maven"
```

The principal difference for me seems to be on the maven cache, but I don't know if this is the cause of the time difference.
I have updated the action version only because of nodejs16 deprecation.

V3:
![setup-java-v3](https://github.com/actions/setup-java/assets/2013524/6ca4a232-a0af-43fc-a74b-df483b7b8f26)

V4:
![setup-java-v4](https://github.com/actions/setup-java/assets/2013524/f7b93de6-ba5f-4895-a932-36e39bb9f9bd)

**Comment by HarithaVattikuti** [comment]
Hello @kdkdhoho 
Thank you for creating this issue. We will investigate it and get back to you as soon as we have some feedback.

**Comment by aparnajyothi-y** [comment]
Hello @brenoavm, Thank you once again for creating this issue.
We have investigated about the issue and able to reproduce the issue but there is no major changes in setup-java that affect the performance time from v3 to v4 other than node 20 upgrade. We tried multiple scenarios and found that the this long time duration is due to the dependencies installation time for node20. As per the node.js official documentation for node 20, we found that from node v19, Global instance of Agent uses keep-alive by default for all HTTP client requests .And the current usage is - HttpClient implementation uses globalAgent when RequestOptions.keepAlive is set to false.
Thank you @fniephaus for this [reference](https://github.com/actions/toolkit/pull/1572). The issue resolves after RequestOptions.keepAlive is applied properly on node20 runtime. Please find the screenshot for reference.

<img width=""971"" alt=""image"" src=""https://github.com/actions/setup-java/assets/147696841/680d2ced-2c58-469f-ba83-6d35b983c92d"">

Please confirm to proceed to close this issue as it is not related to setup-java. Thank you.

**Comment by djthornton1212** [comment]
@aparnajyothi-y, my investigations as to why this post action now takes an extra 2 minutes to run have led me to this issue. 

Like @fniephaus I too am wondering why no one has even reviewed https://github.com/actions/toolkit/pull/1572. There have been 30 something PRs closed since it was opened in October of 2023. I get that there's a lot to be done & it's busy over there, but this is something that's costing companies like my own extra $$$ due to a potential misconfiguration in GitHub's toolkit. 

I know 2 minutes may not sound like a lot, but when you're running 50+ builds a day which use this feature, it adds up quick.

**Comment by fniephaus** [comment]
Could we get an update on this please? In the meantime, a [new release was published](https://github.com/actions/setup-java/releases/tag/v4.1.0) and it's very unfortunate a fix for this was not included.

**Comment by aparnajyothi-y** [comment]
Hello @fniephaus, The [PR](https://github.com/actions/toolkit/pull/1572) merged and waiting for the release to fix the slow post action issue. We will update you once the changes released. Thank you.

**Comment by aparnajyothi-y** [comment]
Hello everyone, The [toolkit PR](https://github.com/actions/toolkit/pull/1572) release and the updated actions/http-client version to 2.2.1 as part of setup-java v4.2.0 release and we tested, the time difference in the post run process in v4. Please check and confirm to proceed to close this issue.

**Comment by fniephaus** [comment]
I can confirm that this is now fixed with `v4.2.0`. Our builds run 3min faster again. Thank you!

**Comment by aparnajyothi-y** [comment]
Hello everyone, Closing this issue as the issue resolved and confirmed.
